contrib/slurm/slurmexecutor.py

```diff
@@ -251,57 +283,120 @@ def run_slurm(benchmark, args, log_file):
                     benchmark.config.singularity,
                 ]
             )
-        srun_command.extend(args)
+        srun_command.extend(
+            [
+                "bash",
+                "-c",
+                f"{' '.join(args)} && echo 0 > exitcode || echo $? > exitcode",
```

Comment:
This would re-introduce all the escaping problems that we eliminated in the last PR. Is there really no way around a shell invocation? Avoiding shell in these situations is really strongly recommended.

Furthermore, in the past we really had a lot of problems with shells mangling the exit code of the benchmarked process, especially if the process terminates due to a signal.

If we need the shell, we must use proper quoting to safely transport all characters such as whitespaces, quotes, `$`, etc.

I would also question whether we really need `bash` or whether we can be more flexible and use any standard shell available via `sh`?

Why `echo 0 > exitcode || echo $? > exitcode` instead of `echo $? > exitcode`?

Reply:
Unfortunately, because SLURM only reports one exitcode (which is 0 if both `srun` and its subcommand returned 0, non-null if _either_ returned non-zero, but sometimes also 0, together with `status: FAILED`, if `srun` experienced a failure), we have to save the subcommand's exitcode somehow. If we don't use a shell, then the only way to retrieve is through the spawning process (in this case, the scheduler), which does not cooperate with us on this. I think we must use a shell. Please correct me if I'm wrong on this, I'd love to go back to using no shells.

How do you suggest I go about quoting the arguments? Are simple `' '` quotes enough, with all single quotes in the arguments being escaped? Is there a utility for this already in Benchexec, or should I try and implement something here?

>   Why `echo 0 > exitcode || echo $? > exitcode` instead of `echo $? > exitcode`?
I'll use a `{command}; echo $? > exitcode` instead, that is cleaner. In the current state, either side would run based on the failure of the command -- the left side (right after `&&`) in the case of a successful run with 0 exitcode, or the right side (after `||`) in the case of a failure. But using `;` instead is cleaner.

Reply:
Is there no way to get the tool's exit code out of the run metadata? For example, the same way you get the CPU time etc.?

If not, I also do not see a way anymore to circumvent the shell.

Reply:
Unfortunately, no. A sample output:

```
Cluster: komondor
User/Group: p_benbl/p_benbl
State: TIMEOUT (exit code 0)
Nodes: 1
Cores per node: 3
CPU Utilized: 00:00:00
CPU Efficiency: 0% of 00:03:09 core-walltime
Job Wall-clock time: 00:01:03
Memory Utilized: 460.39 MB
Memory Efficiency: 9.21% of 4.88 GB
```

Here, the exit code is the combined exit code of the `srun` command as well as the wrapped command.

Reply:
In this example as well, the exit code is supposed to be 0, but the verification task wasn't even finished. For 0 exit codes, we can rely on the status ("COMPLETED" with 0 exit code is when everything executed correctly), but for non-zero exit codes, there is no way of distinguishing between an internal `srun` error and an error during tool execution.

Reply:
They should add this - like `runexec` does ;-)

So maybe switch to `/bin/sh`? `$?` should be available in all shells, not just bash.

Reply:
Sure, I'll do this and run a quick test.

And I agree, just to think how much of the headache could have been avoided, only if SLURM relied on something akin to runexec... :)